zetaclient/maintenance/shutdown_listener.go

```diff
@@ -0,0 +1,125 @@
+package maintenance
+
+import (
+	"context"
+	"time"
+
+	"cosmossdk.io/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/zeta-chain/node/pkg/bg"
+	"github.com/zeta-chain/node/pkg/retry"
+	observertypes "github.com/zeta-chain/node/x/observer/types"
+	"github.com/zeta-chain/node/zetaclient/chains/interfaces"
+)
+
+const restartListenerTicker = 10 * time.Second
+
+// ShutdownListener is a struct that listens for scheduled shutdown notices via the observer
+// operational flags
+type ShutdownListener struct {
+	client interfaces.ZetacoreClient
+	logger zerolog.Logger
+
+	lastRestartHeightMissed int64
+}
+
+// NewShutdownListener creates a new ShutdownListener.
+func NewShutdownListener(client interfaces.ZetacoreClient, logger zerolog.Logger) *ShutdownListener {
+	log := logger.With().Str("module", "shutdown_listener").Logger()
+	return &ShutdownListener{
+		client: client,
+		logger: log,
+	}
+}
+
+func (o *ShutdownListener) Listen(ctx context.Context, action func()) {
+	var (
+		withLogger = bg.WithLogger(o.logger)
+		onComplete = bg.OnComplete(action)
+	)
+
+	bg.Work(ctx, o.waitForUpdate, bg.WithName("shutdown_listener.wait_for_update"), withLogger, onComplete)
+}
+
+func (o *ShutdownListener) waitForUpdate(ctx context.Context) error {
```

Comment:
consider using `ticker` package as it contains exactly the same logic but makes code shorter

Reply:
With `ticker` I'd return a sentinel error when we want to signal for shutdown right? I'm not sure I like that.

Reply:
it will return `nil`. Example: https://github.com/zeta-chain/node/blob/cfcf706956f9e01473a2e781c1bacdb44f70a938/zetaclient/orchestrator/contextupdater.go#L39

Reply:
I will consider refactoring this in one of the next PRs